### ISSUE TITLE
Using `:write` after any command that produces changes introduces a phantom change

### ISSUE THREAD

**Issue by paveloom** [helix-editor/helix#2192]
### Summary

Something in the code introduces a change of state after executing a sequence of commands that contains `:pipe` and ends with `:write`.

### Reproduction Steps

Consider the following config:

```toml
# ~/.config/helix/config.toml
# <...>
[keys.normal]
C-l = [":pipe echo '# test'", ":write"]
# <...>
```

After restarting Helix, do the following:
1. Open any writable buffer
2. Press <kbd>Ctrl + l</kdb>
3. Quit the editor, aborting the changes: `:q!`
4. Restart, open the same buffer, discover that the text was actually written. 

*I expected this to happen:*
The key should print `# test` and write the buffer. 

*Instead, this happened:*
The key does the above, but it also introduces a phantom change. That is, the state of the buffer appears to be modified, according to the indicator in the status bar.

### Helix log

Reproducing this issue doesn't emit relevant log messages.

### Platform

Linux

### Terminal Emulator

GNOME Terminal 3.42.2 using VTE 0.66.2 +BIDI +GNUTLS +ICU +SYSTEMD

### Helix Version

helix 22.05-dev (35d26936)

**Comment by jappeace** [comment]
oh `:pipe sleep 0.1` also produces a change according to the buffer.

so here you got two bugs, 1 pipe producing a change even though one didn't neccisarly happen, 2 write not clearing the change if a change happened

**Comment by Paveloom** [comment]
Correction: the issue happens with any command that produces changes. Try

```toml
# ~/.config/helix/config.toml
# <...>
[keys.normal]
C-h = ["indent", ":write"]
# <...>
```

@jappeace I would argue that `:pipe sleep 0.1` producing a change is expected, since the change actually happens. According to the code, `:pipe` will replace all current selections with the output of a shell command, which in this case is empty.

**Comment by Paveloom** [comment]
So... here's a chain of relevant calls on `write`:

https://github.com/helix-editor/helix/blob/d2b1add1f41c9d9a655a4637e3e276eb92e8eefb/helix-term/src/commands/typed.rs#L235-L241

In `write_impl`:
https://github.com/helix-editor/helix/blob/d2b1add1f41c9d9a655a4637e3e276eb92e8eefb/helix-term/src/commands/typed.rs#L224-L225

https://github.com/helix-editor/helix/blob/d2b1add1f41c9d9a655a4637e3e276eb92e8eefb/helix-view/src/document.rs#L441-L447

In `save_impl`:
https://github.com/helix-editor/helix/blob/d2b1add1f41c9d9a655a4637e3e276eb92e8eefb/helix-view/src/document.rs#L469-L470

https://github.com/helix-editor/helix/blob/d2b1add1f41c9d9a655a4637e3e276eb92e8eefb/helix-view/src/document.rs#L832-L838

So, `reset_modified` should save the current revision as the last one, therefore marking the state as not modified. However, it's still modified after running a sequence of commands because `self.changes` isn't empty, so the `is_modified` method will return `true`:

https://github.com/helix-editor/helix/blob/d2b1add1f41c9d9a655a4637e3e276eb92e8eefb/helix-view/src/document.rs#L824-L830

Taking the example from above, the difference between the mapped sequence of commands `["indent", ":write"]` and two manual calls to `>` and `:w` is that in the latter case, the changes are appended to the history *before* they're written:

In `handle_event`:
https://github.com/helix-editor/helix/blob/d2b1add1f41c9d9a655a4637e3e276eb92e8eefb/helix-term/src/ui/editor.rs#L1255-L1259

https://github.com/helix-editor/helix/blob/d2b1add1f41c9d9a655a4637e3e276eb92e8eefb/helix-view/src/document.rs#L799-L818

A solution to this issue would be to append the changes of *each* command in a command sequence to history. Or maybe only before commands that *should* mark the document as unmodified. Speaking of which, there is a [new](https://github.com/helix-editor/helix/pull/2115) handy-dandy command to append the changes manually:

https://github.com/helix-editor/helix/blob/d2b1add1f41c9d9a655a4637e3e276eb92e8eefb/helix-term/src/commands.rs#L3005-L3008

One can use it as a workaround to the original issue as follows:

```toml
# ~/.config/helix/config.toml
# <...>
[keys.normal]
C-l = [":pipe echo '# test'", "commit_undo_checkpoint", ":write"]
C-h = ["indent", "commit_undo_checkpoint", ":write"]
# <...>
```

**Comment by sourcefrog** [comment]
Another place that a similar or the same bug shows up is if you bind `C-s` to `:write` in insert mode: it does save the file, but Helix thinks it's still modified.

**Comment by JakeEP** [comment]
Came across this issue while searching to see if the problem @sourcefrog mentions already has an issue filed. Unsure if this "save in insert mode has remaining [+]" problem is covered by this issue. Either way, hope it gets a fix soon-ish, as while it's a minor visual bug it is quite distracting for me personally.

**Comment by dead10ck** [comment]
This was likely fixed by #2267. Feel free to reopen if not.

**Comment by TobTobXX** [comment]
Hmm... it does not appear fixed (at least on current HEAD: 789833c9953f1639ff4c580de326d3bb7b9210f9).

I have this config.toml
```toml
[keys.insert]
"esc" = ["normal_mode", ":write"]
```

It does work, but it doesn't update the statusline correctly. (Try inserting in a file and then hit escape. The change was written to disk, but the statusline still says it's dirty.)

Interestingly, when I press `i`, don't make a change and press `esc` immediately after, the flag disappears in the statusline.

I'm not 100% sure if this is the bug mentioned here. I'll open another issue if needed.
